lib/utils/compose.js

```diff
@@ -0,0 +1,1129 @@
+
+	let renderer;
+	if (inlineLogs) {
+		renderer = new BuildProgressInline(
+			logger.streams['build'],
+			imageDescriptors,
+		);
+	} else {
+		const tty = require('./tty')(process.stdout);
+		renderer = new BuildProgressUI(tty, imageDescriptors);
+	}
+	renderer.start();
+
+	return Promise.resolve(checkBuildSecretsRequirements(docker, projectPath))
+		.then(() => qemu.installQemuIfNeeded(emulated, logger, arch, docker))
+		.tap(function(needsQemu) {
+			if (!needsQemu) {
+				return;
+			}
+			logger.logInfo('Emulation is enabled');
+			// Copy qemu into all build contexts
+			return Promise.map(imageDescriptors, function(d) {
+				if (typeof d.image === 'string' || d.image.context == null) {
+					return;
+				}
+				// external image
+				return qemu.copyQemu(path.join(projectPath, d.image.context), arch);
+			});
+		})
+		.then((
+			needsQemu, // Tar up the directory, ready for the build stream
+		) =>
+			tarDirectory(projectPath, { convertEol })
+				.then(tarStream =>
+					makeBuildTasks(
+						composition,
+						tarStream,
+						{ arch, deviceType },
+						logger,
+						projectName,
+					),
+				)
+				.map(function(/** @type {any} */ task) {
+					const d = imageDescriptorsByServiceName[task.serviceName];
+
+					// multibuild parses the composition internally so any tags we've
+					// set before are lost; re-assign them here
+					if (task.tag == null) {
+						task.tag = [projectName, task.serviceName].join('_').toLowerCase();
+					}
+					if (typeof d.image !== 'string' && d.image.context != null) {
+						d.image.tag = task.tag;
+					}
+
+					// configure build opts appropriately
+					if (task.dockerOpts == null) {
+						task.dockerOpts = {};
+					}
+					_.merge(task.dockerOpts, buildOpts, { t: task.tag });
+					if (typeof d.image !== 'string') {
+						/** @type {any} */
+						const context = d.image.context;
+						if (context?.args != null) {
+							if (task.dockerOpts.buildargs == null) {
+								task.dockerOpts.buildargs = {};
+							}
+							_.merge(task.dockerOpts.buildargs, context.args);
+						}
+					}
+
+					// Get the service-specific log stream
+					// Caveat: `multibuild.BuildTask` defines no `logStream` property
+					// but it's convenient to store it there; it's JS ultimately.
+					task.logStream = renderer.streams[task.serviceName];
+					task.logBuffer = [];
+
+					// Setup emulation if needed
+					if (task.external || !needsQemu) {
+						return [task, null];
+					}
+					const binPath = qemu.qemuPathInContext(
+						path.join(projectPath, task.context ?? ''),
+					);
+					if (task.buildStream == null) {
+						throw new Error('No buildStream for task');
+					}
+					return transpose
+						.transposeTarStream(
+							task.buildStream,
+							{
+								hostQemuPath: toPosixPath(binPath),
+								containerQemuPath: `/tmp/${qemu.QEMU_BIN_NAME}`,
+								qemuFileMode: 0o555,
+							},
+							dockerfilePath || undefined,
+						)
+						.then((/** @type {any} */ stream) => {
```

Comment:
`task.buildStream` is declared as a `Pack` which is stream-like but this `stream` is a real stream so we do the cast here to allow things to continue to work like they did in coffeescript but should be handled in a better way when converted to typescript